WildbookOrg/ibeis
# Copy Dockerfiles from private dev repo to devops/

- Copy Dockerfiles from private dev repo to devops/

  `devops/Dockerfile`, `devops/base/Dockerfile` and
  `devops/dependencies/Dockerfile` are unchanged and
  `devops/provision/Dockerfile` has been modified to use public repos on
  github instead of private azure ones.
  
  `devops/provision/_config/types_c.h` is required by the provision
  Dockerfile so has also been copied.
  
  `theanorc`, `embed.sh`, `healthcheck.py` and `boto.cfg` in
  `devops/provision/_config/` are required by the ibeis Dockerfile so have
  been copied.  The aws access key id and secret access key have been
  removed from `boto.cfg`.

- Change Dockerfiles to use base images outside of azure

  The docker images on azurecr.io need authentication, so changing them to
  use docker hub or locally built images instead:
  
  - `wildme.azurecr.io/ibeis/base:latest` -> `wildme/ibeis-base:latest`
  - `wildme.azurecr.io/ibeis/dependencies:latest` -> `wildme/ibeis-dependencies:latest`
  - `wildme.azurecr.io/ibeis/provision:latest` -> `wildme/ibeis-provision:latest`

- Add container build script and instructions


- Use the ibeis master branch in Dockerfile

  The "next" branch on the ibeis github repo is now deprecated.  We are
  switching to gitflow and "next" is only around just in case.

- Set LC_ALL and LANG to C.UTF-8 in Dockerfile

  When running `super_setup.py`:
  
  ```
  Traceback (most recent call last):
    File "super_setup.py", line 866, in <module>
      main()
    File "super_setup.py", line 862, in main
      cli_group()
    File "/virtualenv/env3/lib/python3.6/site-packages/click/core.py", line 829, in __call__
      return self.main(*args, **kwargs)
    File "/virtualenv/env3/lib/python3.6/site-packages/click/core.py", line 760, in main
      _verify_python3_env()
    File "/virtualenv/env3/lib/python3.6/site-packages/click/_unicodefun.py", line 130, in _verify_python3_env
      " mitigation steps.{}".format(extra)
  RuntimeError: Click will abort further execution because Python 3 was configured to use ASCII as encoding for the environment. Consult https://click.palletsprojects.com/python3/ for mitigation steps.
  
  This system supports the C.UTF-8 locale which is recommended. You might be able to resolve your issue by exporting the following environment variables:
  
      export LC_ALL=C.UTF-8
      export LANG=C.UTF-8
  ```
  
  So just set the environment variables in the provision Dockerfile.

- Adjust super_setup.py usage in Dockerfile

  The interface for `super_setup.py` has changed so it's just `pull`
  (subcommand) instead of `--pull`:
  
  ```
  Usage: super_setup.py [OPTIONS] COMMAND [ARGS]...
  
  Options:
    -h, --help  Show this message and exit.
  
  Commands:
    check         Check is just a dry run of "ensure".
    develop
    doctest
    ensure        Ensure is the live run of "check".
    ensure_clone
    pull
    status
    versions
  ```

- Temporarily disable super_setup.py in Dockerfiles

  The repos super_setup.py expecting are different from the ones we check
  out, for example "guitool" vs "guitool_ibeis", so just comment out the
  super_setup.py lines in the dockerfiles for now.

- Temporarily install flann v2.0.0 in Dockerfile

  We usually install the "next" branch for all of our github repos.  But
  for flann, the unix_build.sh on the next branch isn't working after
  recent changes, so use the latest tag v2.0.0 instead.
  
  The latest version of flann requires a cmake version 3.13.0 or higher
  but ubuntu 18.04 only has 3.10.2.  We will install a newer cmake version
  in the future.

- Temporarily install hesaff v2.0.0 in Dockerfile

  We usually install the "next" branch for all of our github repos.  But
  for hesaff, the unix_build.sh on the next branch isn't working after
  recent changes, so use the latest tag v2.0.0 instead.
  
  The hesaff "next" branch requires cmake version 3.13.0 or higher but ubuntu
  18.04 only has version 3.10.2.  We are going to install a newer version of
  cmake in the future to fix this.

- Update vtool path to unix_build.sh in Dockerfile

  The path to `unix_build.sh` in vtool has moved in the latest "next"
  branch from the repository root `.` to `./dev`.

- Install scikit-build in provision Dockerfile

  The new `guitool` and `vtool` import `skbuild` from `setup.py` so
  requires `scikit-build` to be installed.

- Activate virtualenv before install in dockerfile

  When installing vtool without activating virtualenv, cmake fails:
  
  ```
      ERROR: Command errored out with exit status 1:
       command: /virtualenv/env3/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/ibeis/vtool/setup.py'"'"'; __file__='"'"'/ibeis/vtool/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps
           cwd: /ibeis/vtool/
      Complete output (117 lines):
      Not searching for unused variables given on the command line.
      CMake Error: CMake was unable to find a build program corresponding to "Ninja".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
      -- Configuring incomplete, errors occurred!
  ...
      ---------------------------
      --------------------------------
      -- Trying "Unix Makefiles" generator - success
      --------------------------------------------------------------------------------
  
      Configuring Project
        Working directory:
          /ibeis/vtool/_skbuild/linux-x86_64-3.6/cmake-build
        Command:
          cmake /ibeis/vtool -G 'Unix Makefiles' -DCMAKE_INSTALL_PREFIX:PATH=/ibeis/vtool/_skbuild/linux-x86_64-3.6/cmake-install -DPYTHON_EXECUTABLE:FILEPATH=/virtualenv/env3/bin/python3 -DPYTHON_VERSION_STRING:STRING=3.6.9 -DPYTHON_INCLUDE_DIR:PATH=/usr/include/python3.6m -DPYTHON_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libpython3.6m.so -DSKBUILD:BOOL=TRUE -DCMAKE_MODULE_PATH:PATH=/virtualenv/env3/lib/python3.6/site-packages/skbuild/resources/cmake -DCMAKE_BUILD_TYPE:STRING=Release
  
      Traceback (most recent call last):
  
      ----------------------------------------
    Rolling back uninstall of vtool-ibeis
    Moving to /virtualenv/env3/lib/python3.6/site-packages/vtool_ibeis-2.1.1.dist-info/
     from /virtualenv/env3/lib/python3.6/site-packages/~tool_ibeis-2.1.1.dist-info
    Moving to /virtualenv/env3/lib/python3.6/site-packages/vtool_ibeis/
     from /virtualenv/env3/lib/python3.6/site-packages/~tool_ibeis
  ERROR: Command errored out with exit status 1: /virtualenv/env3/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/ibeis/vtool/setup.py'"'"'; __file__='"'"'/ibeis/vtool/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.
  ```
  
  Activating the virtualenv before installing fixes this problem.
